### DISCUSSION TITLE
How can I get ALL public repos I have contributed to?

### DISCUSSION THREAD

**Discussion by jonathancross** [community/community#24350] [Apps, API and Webhooks] [answered]
<p>I tried using this GraphQL, but it only returns about 30 results with no further pagination:</p>
<pre><code>{
  viewer {
    repositoriesContributedTo(first: 100, contributionTypes: [COMMIT, ISSUE, PULL_REQUEST, REPOSITORY]) {
      nodes {
        nameWithOwner
      }
      pageInfo {
        endCursor
        hasNextPage
      }
    }
  }
}
</code></pre>
<p>I have checked that several normal, public contributions are missing.  I expect a <a href="https://github.com/search?o=desc&amp;q=author%3Ajonathancross&amp;s=author-date&amp;type=Commits&amp;utf8=%E2%9C%93" title="List of Contributions via search" rel="nofollow noopener">much longer list</a>.</p>

**Comment by lee-dohm** [comment]
<p>The documentation of that connection states that it is: “A list of repositories that the user <strong>recently</strong> contributed to.” (emphasis mine) If you want a list of everything, you would have to use the search functionality and create the list of public repositories from the various results types you retrieved from the search APIs.</p>
<p>I hope that helps!</p>

**Comment by dione2** [comment]
<p>Other ways to get help would me git- hub enterprises</p>

**Comment by saoudrizwan** [comment, marked as answer]
<p>Can you please give some more information about what ‘recently’ entails? Is there a time period that GitHub uses to determine ‘recently’? Also, regarding your suggestion about using the search functionality, can you pls give some more details about the implementation you had in mind for retrieving <strong>every</strong> repository a user has contributed to <em>since the beginning of time</em>? I’m wary of abusing the search functionality if I implement it the wrong way for doing something as lightweight as getting a list of repositories. </p>
<p>Thank you!</p>
<p><strong>Edit:</strong> I played around with User.contributionsCollection, but found that it only accepts 1-year long intervals, which would mean I’d have to make as many requests as the number of years the user has made contributions (using contributionsCollection.hasActivityInThePast), or at most ~11 requests (since GitHub was founded in 2008). </p>

**Comment by lee-dohm** [comment]
<p>Unfortunately, no, I don’t have any other information beyond “recently”. Whatever the window is, whether it is a span of time, the most recent X results, or some other metric, it isn’t “everything”.</p>
<p>As for being wary of abusing the search functionality, working within the rate limits is a good first step :grinning: But it sounds like you’ve got a good solution there with <code>User.contributionsCollection</code>. Let us know how it works out!</p>

**Comment by vikrant-sahu** [comment]
<p>Hey…I was looking for the same. Did u get any other method to get all the contributions of a user in one go? </p>

**Comment by SorinGabriel02** [comment]
<p>Interested in the about the same thing, I’m trying to find the most starred repository an user has contributed to (from the beginning of the account)…</p>

**Comment by rickstaa** [comment]
I just created a feature request to improve the user experience. You can show your support at https://github.com/community/community/discussions/39589.

**Comment by rickstaa** [comment]
I just received the following response from GitHub support. The **recently** is defined as one year and the only way to get the total `repositoriesContributedTo` is via the `contributionsCollection` object.

***
Thank you for contacting GitHub Support!

> Can you maybe tell me what recently means? Is this one year, is this the current year?

The repositoriesContributedTo connection returns "a list of repositories that the user recently contributed to," and the "recently" here is defined as "in the last year".

> Further, do you maybe know another way to get all the repositoriesContributedTo a user has contributed to since GitHub was launched? I tried to use the contributionsCollection.totalRepositoriesWithContributedCommits object, but this object also includes a user's own repositories. I also searched the GitHub Rest API, but it also doesn't seem to provide a way.

There isn't a way to get all contributions since GitHub was launched using the repositoriesContributedTo connection. I'd say the contributionsCollection connection is still the way to go, you'd need to filter out the user's own repositories as you noted [in the discussion here](https://github.com/orgs/community/discussions/39589).

I have marked this ticket also as a feedback/feature request so it is visible to the Product team.

Please note that while the Product team reads and evaluates all feedback, we cannot guarantee a response to every submission. We invite you to share your feedback in our [official GitHub community discussions repository](https://github.com/community/community/discussions).

**Comment by yndajas** [comment]
I just used the following to get some stats/data along these lines, albeit year by year, which has already been highlighted as a limitation

```graphql
# Run at https://docs.github.com/en/graphql/overview/explorer

{
  viewer {
    contributionsCollection(
      from: "2024-01-01T00:00:00Z"
      to: "2024-12-31T23:59:59Z"
    ) {
      startedAt
      endedAt
      totalCommitContributions
      totalRepositoriesWithContributedCommits
      # if the `totalRepositoriesWithContributedCommits` number is greater than
      # 100, the `maxRepositories` argument will need increasing, or the query
      # will need splitting
      commitContributionsByRepository(maxRepositories: 100) {
        repository {
          name
          description
          owner {
            ... on Organization {
              name
              resourcePath
              url
            }
            ... on User {
              name
              resourcePath
              url
            }
          }
          resourcePath
          url
        }
        contributions {
          totalCount
        }
        url
      }
      totalIssueContributions
      totalRepositoriesWithContributedIssues
      totalPullRequestContributions
      totalRepositoriesWithContributedPullRequests
      totalPullRequestReviewContributions
      totalRepositoriesWithContributedPullRequestReviews
      totalRepositoryContributions
      contributionCalendar {
        totalContributions
      }
      hasActivityInThePast
    }
  }
}
```

**Reply by vincerubinetti** [reply]
I just wanted to get some stats about myself for a personal portfolio. Here's a small ts script I wrote to gather per-repo stats like commits/prs/etc:

```typescript
import { writeFileSync } from "fs";
import { Octokit } from "@octokit/core";
import { throttling } from "@octokit/plugin-throttling";

const login = "vincerubinetti";
const start = 2008;
const end = new Date().getFullYear();
const output = "repos.json";

const OctokitWithPlugins = Octokit.plugin(throttling);

const octokit = new OctokitWithPlugins({
  auth: process.env.GITHUB_TOKEN,
  throttle: { onRateLimit: () => true, onSecondaryRateLimit: () => true },
});

type Response = {
  user: { contributionsCollection: ContributionsCollection };
};

type ContributionsCollection = {
  commitContributionsByRepository: ContributionsByRepository[];
  issueContributionsByRepository: ContributionsByRepository[];
  pullRequestContributionsByRepository: ContributionsByRepository[];
  pullRequestReviewContributionsByRepository: ContributionsByRepository[];
  totalCommitContributions: number;
  totalIssueContributions: number;
  totalPullRequestContributions: number;
  totalPullRequestReviewContributions: number;
  restrictedContributionsCount: number;
};

type ContributionsByRepository = {
  repository: { nameWithOwner: string };
  contributions: { totalCount: number };
};

const getContributionsForYear = async (login: string, year: number) => {
  const from = `${year}-01-01T00:00:00Z`;
  const to = `${year}-12-31T23:59:59Z`;

  const query = `
    query($login: String!, $from: DateTime!, $to: DateTime!) {
      user(login: $login) {
        contributionsCollection(from: $from, to: $to) {
          commitContributionsByRepository(maxRepositories: 100) {
            repository {
              nameWithOwner
            }
            contributions {
              totalCount
            }
          }
          issueContributionsByRepository(maxRepositories: 100) {
            repository {
              nameWithOwner
            }
            contributions {
              totalCount
            }
          }
          pullRequestContributionsByRepository(maxRepositories: 100) {
            repository {
              nameWithOwner
            }
            contributions {
              totalCount
            }
          }
          pullRequestReviewContributionsByRepository(maxRepositories: 100) {
            repository {
              nameWithOwner
            }
            contributions {
              totalCount
            }
          }
          totalCommitContributions
          totalIssueContributions
          totalPullRequestContributions
          totalPullRequestReviewContributions
          restrictedContributionsCount
        }
      }
    }
  `;

  const response = await octokit.graphql<Response>(query, { login, from, to });
  return response.user.contributionsCollection;
};

type ContributionCounts = {
  commits: number;
  issues: number;
  prs: number;
  reviews: number;
  private: number;
};

type Contributions = {
  [key: string]: ContributionCounts;
  total: ContributionCounts;
};

const counts = { commits: 0, issues: 0, prs: 0, reviews: 0, private: 0 };

const contributions: Contributions = { total: { ...counts } };

const keys: [keyof ContributionsCollection, keyof ContributionCounts][] = [
  ["commitContributionsByRepository", "commits"],
  ["issueContributionsByRepository", "issues"],
  ["pullRequestContributionsByRepository", "prs"],
  ["pullRequestReviewContributionsByRepository", "reviews"],
  ["totalCommitContributions", "commits"],
  ["totalIssueContributions", "issues"],
  ["totalPullRequestContributions", "prs"],
  ["totalPullRequestReviewContributions", "reviews"],
  ["restrictedContributionsCount", "private"],
];

for (let year = start; year < end; year++) {
  const data = await getContributionsForYear(login, year);

  for (const [source, target] of keys) {
    const datum = data[source];
    if (Array.isArray(datum)) {
      for (const repo of datum) {
        const key = repo.repository.nameWithOwner;
        contributions[key] ??= { ...counts };
        contributions[key][target] += repo.contributions.totalCount;
      }
    } else if (typeof datum === "number") contributions.total[target] += datum;
  }
}

writeFileSync(output, JSON.stringify(contributions, null, 2));
```

**Comment by aK128180** [comment]
> <p>Can you please give some more information about what ‘recently’ entails? Is there a time period that GitHub uses to determine ‘recently’? Also, regarding your suggestion about using the search functionality, can you pls give some more details about the implementation you had in mind for retrieving <strong>every</strong> repository a user has contributed to <em>since the beginning of time</em>? I’m wary of abusing the search functionality if I implement it the wrong way for doing something as lightweight as getting a list of repositories. </p>
> <p>Thank you!</p>
> <p><strong>Edit:</strong> I played around with User.contributionsCollection, but found that it only accepts 1-year long intervals, which would mean I’d have to make as many requests as the number of years the user has made contributions (using contributionsCollection.hasActivityInThePast), or at most ~11 requests (since GitHub was founded in 2008). </p>
